lib/ex_unit/lib/ex_unit/assertions.ex

```diff
@@ -357,6 +362,14 @@ defmodule ExUnit.Assertions do
   end
 
   @doc false
+  def __match__({:when, _, _} = left, _, _, _, _) do
+    raise ArgumentError, """
```

Comment:
I'm not sure what the best way to raise here would be.
Ideally this would be a CompileError.

BTW, I noticed that right now without assert we're getting:

```elixir
error: undefined function when/2 (there is no such import)
```

But I think we can have a better error message in this case, given that `when` is a reserved word and can't be used as a function name?